internal/repos/jvm_packages.go

```
@@ -0,0 +1,159 @@
+package repos
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/inconshreveable/log15"
+
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/conf/reposource"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/jvmpackages"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/jvmpackages/coursier"
+	"github.com/sourcegraph/sourcegraph/internal/jsonc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+// A JvmPackagesSource yields depots from a single Maven connection configured
+// in Sourcegraph via the external services configuration.
+type JvmPackagesSource struct {
+	svc    *types.ExternalService
+	config *schema.JvmPackagesConnection
+}
+
+// NewJvmPackagesSource returns a new MavenSource from the given external
+// service.
+func NewJvmPackagesSource(svc *types.ExternalService) (*JvmPackagesSource, error) {
+	var c schema.JvmPackagesConnection
+	if err := jsonc.Unmarshal(svc.Config, &c); err != nil {
+		return nil, fmt.Errorf("external service id=%d config error: %s", svc.ID, err)
+	}
+	return newJvmPackagesSource(svc, &c)
+}
+
+func newJvmPackagesSource(svc *types.ExternalService, c *schema.JvmPackagesConnection) (*JvmPackagesSource, error) {
+	return &JvmPackagesSource{
+		svc:    svc,
+		config: c,
+	}, nil
+}
+
+// ListRepos returns all Maven artifacts accessible to all connections
+// configured in Sourcegraph via the external services configuration.
+func (s JvmPackagesSource) ListRepos(ctx context.Context, results chan SourceResult) {
+	s.listDependentRepos(ctx, results)
+}
+
+func (s JvmPackagesSource) listDependentRepos(ctx context.Context, results chan SourceResult) {
+	modules, err := MavenModules(*s.config)
+	if err != nil {
+		log15.Error("failed to parse JVM modules", "configuration", *s.config, "error", err)
+		return
```

Do you not want to return the error as `SourcdeResult{Error: err}` on the results chan?

---

TIL that's much cleaner